File: sol/lab10/cpp/task-1/sol2_mst_prim.cpp
// SPDX-License-Identifier: BSD-3-Clause

#include <algorithm>
#include <fstream>
#include <iostream>
#include <queue>
#include <set>
#include <utility>
#include <vector>
using namespace std;

// numarul maxim de noduri
#define NMAX 200005

// valoare mai mare decat orice distanta din graf
#define INF (1 << 30)

struct Edge {
    int node;
    int neigh;
    int w;

    Edge() { }
    Edge(int node, int neigh, int w)
        : node(node)
        , neigh(neigh)
        , w(w) { }
};

// structura folosita pentru a stoca MST
struct MSTResult {
    int cost; // costul MST-ului gasit

    vector<pair<int, int>> edges; // muchiile din MST-ul gasit (ordinea nu conteaza)

    MSTResult(int cost, const vector<pair<int, int>>& edges)
        : cost(cost)
        , edges(edges) { }
};

class Task {
public:
    void solve() {
        read_input();
        print_output(get_result());
    }

private:
    // n = numar de noduri, m = numar de muchii
    int n, m;

    // muchiile din graf: (node, neigh, w) - muchie de la node la neigh de cost w
    vector<Edge> edges;

    void read_input() {
        ifstream fin("in");
        fin >> n >> m;
        for (int i = 1, x, y, w; i <= m; i++) {
            fin >> x >> y >> w;
            edges.push_back(Edge{x, y, w});
        }
        fin.close();
    }

    MSTResult get_result() {
        // TODO: Calculati costul minim al unui MST folosind algoritmul lui Prim.
        // Alegem pentru simplitate nodul de start 1.
        return prim(n, edges);
    }

    // MST generat cu Prim, pornind de la un nod radacina ales anterior.
    // Complexitate: O(m log n).
    MSTResult prim(int nodes, vector<Edge>& edges, int source = 1) {
        // adj[node] = lista de adiacenta a lui node: (neigh, w)
        vector<pair<int, int>> adj[NMAX];
        for (const auto& [x, y, w] : edges) {
            adj[x].push_back({y, w});
            adj[y].push_back({x, w});
        }

        // d[node] = distanta nodului node fata de MST-ul curent (cel ma apropiat nod din MST)
        // d[node] in aceasta problema va fi mereu egal cu costul unei muchii
        // Initializam vectorul de distante cu distante infinite.
        vector<int> d(n + 1, INF);
        // p[node] = parintele lui node (initializat cu 0)
        vector<int> p(n + 1, 0);
        vector<int> used(n + 1, 0);

        // In loc de un priority_queue<> folosim set<>, cu elemente tupluri de
        // tipul (distanta pana la nod, nod). Setul tine elementele ordonate, deci
        // mereum in pq.begin() vom avea tuplul care ii corespunde nodului cu
        // distanta minima.
        set<pair<int, int>> pq;

        // Inseram nodul de plecare in coada si ii actualizam distanta.
        d[source] = 0;
        p[source] = 0;
        pq.insert({d[source], source});

        // Initializam MST: cost 0, fara muchii
        int cost = 0;
        vector<pair<int, int>> mst;

        // Adaugam fix n noduri la arbore
        while (!pq.empty()) {
            // Scoatem un nod din coada.
            int node = pq.begin()->second;
            pq.erase(pq.begin());
            // In cazul in care nodul e deja in MST, ignoram aceasta intrare.
            if (used[node]) {
                continue;
            }

            // Adaug muchia node - p[node].
            used[node] = 1;
            // Nodul radacina este adaugat print-o muchie fictiva,
            // care nu face parte din MST.
            if (p[node]) {
                cost += d[node];
                mst.push_back({node, p[node]});
            }

            // Ii parcurgem toti vecinii.
            for (const auto& [neigh, w] : adj[node]) {
                // Se imbunatateste distanta?
                if (!used[neigh] && w < d[neigh]) {
                    // Scoatem vechea pereche din set.
                    pq.erase({d[neigh], neigh});
                    // Actualizam distanta si inseram din nou in set.
                    d[neigh] = w;
                    p[neigh] = node;
                    pq.insert({d[neigh], neigh});
                }
            }
        }

        return {cost, mst};
    }

    void print_output(const MSTResult& res) {
        ofstream fout("out");
        fout << res.cost << "\n";
        for (const auto& [x, y] : res.edges) {
            fout << x << " " << y << "\n";
        }
        fout.close();
    }
};

// [ATENTIE] NU modifica functia main!
int main() {
    // * se aloca un obiect Task pe heap
    // (se presupune ca e prea mare pentru a fi alocat pe stiva)
    // * se apeleaza metoda solve()
    // (citire, rezolvare, printare)
    // * se distruge obiectul si se elibereaza memoria
    auto* task = new (nothrow) Task(); // hint: cppreference/nothrow
    if (!task) {
        cerr << "new failed: WTF are you doing? Throw your PC!\n";
        return -1;
    }
    task->solve();
    delete task;
    return 0;
}

File: sol/lab10/java/task-1/src/sol2_mst_prim.java
// SPDX-License-Identifier: BSD-3-Clause

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.PriorityQueue;

public class sol2_mst_prim {
    static class Task {
        public static final String INPUT_FILE = "in";
        public static final String OUTPUT_FILE = "out";

        // numarul maxim de noduri
        public static final int NMAX = 200005;

        // valoare mai mare decat orice distanta din graf
        public static final int INF = (1 << 30);

        // n = numar de noduri, m = numar de muchii
        int n, m;

        // adj[node] = lista de adiacenta a nodului node
        // Edge e inseamna ca exista muchie de la e.node la e.neigh de cost e.w
        ArrayList<Edge> edges = new ArrayList<>();

        public class Pair {
            public int x;
            public int y;

            Pair(int _x, int _y) {
                x = _x;
                y = _y;
            }
        }

        public class Edge {
            int node;
            int neigh;
            int w;

            Edge(int _node, int _neigh, int _w) {
                node = _node;
                neigh = _neigh;
                w = _w;
            }
        };

        // structura folosita pentru a stoca MST
        class MSTResult {
            int cost; // costul MST-ului gasit

            ArrayList<Pair> edges; // muchiile din MST-ul gasit (ordinea nu conteaza)

            MSTResult(int _cost,  ArrayList<Pair> _edges) {
                cost = _cost;
                edges = _edges;
            }
        };

        public void solve() {
            readInput();
            writeOutput(getResult());
        }

        private void readInput() {
            try {
                Scanner sc = new Scanner(new BufferedReader(new FileReader(
                                INPUT_FILE)));
                n = sc.nextInt();
                m = sc.nextInt();

                for (int i = 1; i <= m; i++) {
                    int x, y, w;
                    x = sc.nextInt();
                    y = sc.nextInt();
                    w = sc.nextInt();
                    edges.add(new Edge(x, y, w));
                }
                sc.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private void writeOutput(MSTResult result) {
            try {
                PrintWriter pw = new PrintWriter(new File(OUTPUT_FILE));
                pw.printf("%d\n", result.cost);
                for (Pair e : result.edges) {
                    pw.printf("%d %d\n", e.x, e.y);
                }
                pw.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private MSTResult getResult() {
            // TODO: Calculati costul minim al unui MST folosind algoritmul lui Prim.
            // Alegem pentru simplitate nodul de start 1.
            return prim(n, edges, 1);
        }

        
        public class PairPrim implements Comparable<PairPrim> {
            public int destination;
            public int cost;

            PairPrim(int _destination, int _cost) {
                destination = _destination;
                cost = _cost;
            }

            public int compareTo(PairPrim rhs) {
                return Integer.compare(cost, rhs.cost);
            }
        }

        // MST generat cu Prim, pornind de la un nod radacina ales anterior.
        // Complexitate: O(m log n).
        private MSTResult prim(int nodes, ArrayList<Edge> edges, int source) {
            // adj[node] = lista de adiacenta a lui node: (neigh, w)
            ArrayList<PairPrim> adj[] = new ArrayList[NMAX];
            for (int node = 1; node <= nodes; ++node) {
                adj[node] = new ArrayList<>();
            }
            for (Edge e : edges) {
                adj[e.node].add(new PairPrim(e.neigh, e.w));
                adj[e.neigh].add(new PairPrim(e.node, e.w));
            }

            // d[node] = distanta nodului node fata de MST-ul curent (cel ma apropiat nod din MST)
            // d[node] in aceasta problema va fi mereu egal cu costul unei muchii
            // Initializam vectorul de distante cu distante infinite.
            int[] d = new int[n + 1];
            // p[node] = parintele lui node (initializat cu 0)
            int[] p = new int[n + 1];
            boolean[] used = new boolean[n + 1];

            // Initializam distantele la infinit.
            for (int node = 0; node <= nodes; node++) {
                d[node] = INF;
                p[node] = 0;
            }

            // Folosim un priority queue de PairPrim, desi elementele noastre nu sunt tocmai muchii.
            // Vom folosi field-ul de cost ca si distanta pana la nodul respectiv.
            // Observati ca am modificat clasa Edge ca sa implementeze interfata Comparable.
            PriorityQueue<PairPrim> pq = new PriorityQueue<>();

            // Inseram nodul de plecare in coada si ii setam distanta pe 0.
            d[source] = 0;
            pq.add(new PairPrim(source, 0));

            // Initializam MST: cost 0, fara muchii
            int cost = 0;
            ArrayList<Pair> mst = new ArrayList<>();

            // Cat timp mai sunt noduri in coada
            while (!pq.isEmpty()) {
                // Scoatem un nod din coada
                int node = pq.poll().destination;
                // In cazul in care nodul e deja in MST, ignoram aceasta intrare.
                if (used[node]) {
                    continue;
                }

                // Adaug muchia node - p[node].
                used[node] = true;
                // Nodul radacina este adaugat print-o muchie fictiva,
                // care nu face parte din MST.
                if (p[node] != 0) {
                    cost += d[node];
                    mst.add(new Pair(node, p[node]));
                }

                // Ii parcurgem toti vecinii.
                for (PairPrim e : adj[node]) {
                    int neigh = e.destination, w = e.cost;

                    // Se imbunatateste distanta?
                    if (!used[neigh] && w < d[neigh]) {
                        // Actualizam distanta si inseram din nou in pq.
                        d[neigh] = w;
                        p[neigh] = node;
                        pq.add(new PairPrim(neigh, d[neigh]));
                    }
                }
            }

            return new MSTResult(cost, mst);
        }
    }

    public static void main(String[] args) {
        new Task().solve();
    }
}
